src/tiled/moveselectiontool.cpp

```diff
@@ -0,0 +1,197 @@
+
+MoveSelectionTool::MoveSelectionTool(QObject *parent)
+    : AbstractTileTool(tr("Move Selection"),
+                       QIcon(QLatin1String(
+                               ":images/22x22/move-selection.png")),
+                       QKeySequence(tr("G")),
+                       parent)
+    , mDragging(false)
+    , mMouseDown(false)
+    , mCut(false)
+{
+}
+
+MoveSelectionTool::~MoveSelectionTool()
+{
+}
+
+void MoveSelectionTool::tilePositionChanged(const QPoint &pos)
+{
+    if (mDragging) {
+
+        QPoint offset = pos - mLastUpdate;
+        QRegion selectedArea = mapDocument()->selectedArea();
+        selectedArea.translate(offset);
+
+        mapDocument()->undoStack()->push(new ChangeSelectedArea(mapDocument(), selectedArea));
+
+        if (mPreviewLayer) {
+            mPreviewLayer->offsetTiles(offset, mPreviewLayer->bounds(), false, false);
+            brushItem()->setTileLayer(mPreviewLayer);
+        }
+
+        mLastUpdate = pos;
+
+        brushItem()->setTileRegion(mapDocument()->selectedArea());
+    }
+}
+
+void MoveSelectionTool::mouseMoved(const QPointF &pos, Qt::KeyboardModifiers modifiers)
+{
+    if (mMouseDown && !mDragging) {
+        QPoint screenPos = QCursor::pos();
+        const int dragDistance = (mMouseScreenStart - screenPos).manhattanLength();
+
+        if (dragDistance >= QApplication::startDragDistance() / 2) {
+            mDragging = true;
+            tilePositionChanged(tilePosition());
+            if (!mCut) {
+                mCut = true;
+                cut();
+            }
+        }
+    }
+
+    AbstractTileTool::mouseMoved(pos, modifiers);
+
+    refreshCursor();
+}
+
+void MoveSelectionTool::mousePressed(QGraphicsSceneMouseEvent *event)
+{
+    if (event->button() == Qt::LeftButton && mapDocument()->selectedArea().contains(tilePosition())) {
+        mMouseScreenStart = event->screenPos();
+        mDragStart = tilePosition();
+        mMouseDown = true;
+        mLastUpdate = tilePosition();
+    }
+}
+
+void MoveSelectionTool::mouseReleased(QGraphicsSceneMouseEvent *)
+{
+    if (mDragging) {
+        mDragging = false;
+        brushItem()->setTileRegion(QRegion());
+    }
+
+    if (mCut) {
+        paste();
+        mCut = false;
+    }
+
+    mMouseDown = false;
+    refreshCursor();
+}
+
+void MoveSelectionTool::languageChanged()
+{
+    setName(tr("Move Selection"));
+    setShortcut(QKeySequence(tr("G")));
+}
+
+void MoveSelectionTool::refreshCursor()
+{
+    Qt::CursorShape cursorShape = Qt::ArrowCursor;
+
+    if (mDragging)
+        cursorShape = Qt::SizeAllCursor;
+
+    if (cursor().shape() != cursorShape)
+        setCursor(cursorShape);
+}
+
+void MoveSelectionTool::cut()
+{
+    Layer *currentLayer = mapDocument()->currentLayer();
+
+    TileLayer *tileLayer = dynamic_cast<TileLayer*>(currentLayer);
+    const QRegion &selectedArea = mapDocument()->selectedArea();
+
+    ClipboardManager::instance()->copySelection(mapDocument());
```

Comment:
I still think that there should be no `ClipboardManager` involved here... One would not expect to loose the current clipboard content when using the tool. It should be possible to store the moved selection in the tool.